Jadx can not parse resource type chunks with unordered entries

### Issue details

https://www.apkmirror.com/apk/whatsapp-inc/whatsapp/whatsapp-2-24-24-14-release/whatsapp-messenger-2-24-24-14-android-apk-download/

### Relevant log output or stacktrace

```java
ERROR - Failed to parse 'resources.pb'/'.arsc' file
jadx.core.utils.exceptions.JadxException: Error decode: resources.arsc
	at jadx.api.ResourcesLoader.decodeStream(ResourcesLoader.java:118)
	at jadx.core.dex.nodes.RootNode.loadResources(RootNode.java:212)
	at jadx.api.JadxDecompiler.load(JadxDecompiler.java:128)
	at jadx.cli.JadxCLI.processAndSave(JadxCLI.java:59)
	at jadx.cli.JadxCLI.execute(JadxCLI.java:37)
	at jadx.cli.JadxCLI.main(JadxCLI.java:27)
Caused by: java.io.IOException: Expected start of entry 1, expected offset: 0x1f02f4, actual: 0x1f0304
	at jadx.core.xmlgen.ParserStream.checkPos(ParserStream.java:125)
	at jadx.core.xmlgen.ResTableBinaryParser.parseTypeChunk(ResTableBinaryParser.java:289)
	at jadx.core.xmlgen.ResTableBinaryParser.parsePackage(ResTableBinaryParser.java:164)
	at jadx.core.xmlgen.ResTableBinaryParser.decodeTableChunk(ResTableBinaryParser.java:111)
	at jadx.core.xmlgen.ResTableBinaryParser.decode(ResTableBinaryParser.java:84)
	at jadx.api.ResourcesLoader.decodeTable(ResourcesLoader.java:173)
	at jadx.core.dex.nodes.RootNode.lambda$loadResources$7(RootNode.java:212)
	at jadx.api.ResourcesLoader.decodeStream(ResourcesLoader.java:113)
	... 5 common frames omitted
INFO  - processing ...
INFO  - done
```


### Provide sample and class/method full name

_No response_

### Jadx version

1.5.1

--- a/jadx-cli/src/main/java/jadx/cli/tools/ConvertArscFile.java
+++ b/jadx-cli/src/main/java/jadx/cli/tools/ConvertArscFile.java
@@ -1,6 +1,5 @@
 package jadx.cli.tools;
 
-import java.io.BufferedInputStream;
 import java.io.IOException;
 import java.io.InputStream;
 import java.nio.file.Files;
@@ -71,7 +70,7 @@ public class ConvertArscFile {
 				}
 			} else {
 				// Load resources.arsc from extracted file
-				try (InputStream inputStream = new BufferedInputStream(Files.newInputStream(resFile))) {
+				try (InputStream inputStream = Files.newInputStream(resFile)) {
 					resTableParser.decode(inputStream);
 				}
 			}

--- a/jadx-core/src/main/java/jadx/core/xmlgen/ParserStream.java
+++ b/jadx-core/src/main/java/jadx/core/xmlgen/ParserStream.java
@@ -18,6 +18,7 @@ public class ParserStream {
 
 	private final InputStream input;
 	private long readPos = 0;
+	private long markPos = 0;
 
 	public ParserStream(@NotNull InputStream inputStream) {
 		this.input = inputStream;
@@ -143,10 +144,12 @@ public class ParserStream {
 			throw new IOException("Mark not supported for input stream " + input.getClass());
 		}
 		input.mark(len);
+		markPos = readPos;
 	}
 
 	public void reset() throws IOException {
 		input.reset();
+		readPos = markPos;
 	}
 
 	public void readFully(byte[] b) throws IOException {

--- a/jadx-core/src/main/java/jadx/core/xmlgen/ResTableBinaryParser.java
+++ b/jadx-core/src/main/java/jadx/core/xmlgen/ResTableBinaryParser.java
@@ -1,5 +1,6 @@
 package jadx.core.xmlgen;
 
+import java.io.BufferedInputStream;
 import java.io.IOException;
 import java.io.InputStream;
 import java.util.ArrayList;
@@ -79,7 +80,7 @@ public class ResTableBinaryParser extends CommonBinaryParser implements IResTabl
 	@Override
 	public void decode(InputStream inputStream) throws IOException {
 		long start = System.currentTimeMillis();
-		is = new ParserStream(inputStream);
+		is = new ParserStream(new BufferedInputStream(inputStream, 32768));
 		resStorage = new ResourceStorage(root.getArgs().getSecurity());
 		decodeTableChunk();
 		resStorage.finish();
@@ -234,6 +235,7 @@ public class ResTableBinaryParser extends CommonBinaryParser implements IResTabl
 		/* int size = */
 		long chunkSize = is.readUInt32();
 		long chunkEnd = start + chunkSize;
+		is.mark((int) chunkSize);
 
 		// The type identifier this chunk is holding. Type IDs start at 1 (corresponding
 		// to the value of the type bits in a resource identifier). 0 is invalid.
@@ -286,7 +288,12 @@ public class ResTableBinaryParser extends CommonBinaryParser implements IResTabl
 				LOG.warn("End of chunk reached - ignoring remaining {} entries", entryCount - index);
 				break;
 			}
-			is.checkPos(entriesStart + offset, "Expected start of entry " + index);
+			long entryStartOffset = entriesStart + offset;
+			if (entryStartOffset < is.getPos()) {
+				// workaround for issue #2343: if the entryStartOffset is located before our current position
+				is.reset();
+			}
+			is.skipToPos(entryStartOffset, "Expected start of entry " + index);
 			parseEntry(pkg, id, index, config.getQualifiers());
 		}
 		if (chunkEnd > is.getPos()) {

